### DISCUSSION TITLE
Snapshot validation

### DISCUSSION THREAD

**Discussion by KislyakovDS** [mobxjs/mobx-state-tree#1940] [Q&A] [answered]
Is there some best way to validate snapshot for a model? I need to check snapshot for `types.union` dispatcher.

```js
const Node = types.union(
	{ dispatcher: (snapshot) => 'message' in snapshot ? Message : Form },
	Form,
	Message
);
```

I don't want to check properties in snapshot. I have an idea to create model instance with snapshot and check if error will be throw. May be there is best practice for this case?

**Comment by EmilTholin** [comment, marked as answer]
Hi @KislyakovDS!

The `dispatcher` option isn't really a validator, but rather a function to help determine what type of model the snapshot actually represents. [I don't think you will be able to skip this if you have a `union` with two custom models](https://github.com/mobxjs/mobx-state-tree/blob/dafdb9c371a6a908d1e7e097fe2efd3a9a31abaa/packages/mobx-state-tree/src/types/utility-types/union.ts#L93-L114).

You could e.g. do what the [Ignite boilerplate does and try to instantiate a model in a `try`/`catch`, and have some fallback behavior if an error is thrown](https://github.com/infinitered/ignite/blob/bdbec9a2b96898bc78ed48fd072fc268b6662387/boilerplate/app/models/root-store/setup-root-store.ts#L33-L44), but again I don't think this will help with determining the type of the union.

**Reply by KislyakovDS** [reply]
Ok, I understand, thanks a lot. I have more complicated structure of models and I provided simple example.
